src/core/lib/transport/metadata_batch.cc

```diff
@@ -25,11 +25,103 @@
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_CONTENT_LENGTH_EMPTY_INDEX, 14 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_CONTENT_LOCATION_EMPTY_INDEX, 16 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_CONTENT_RANGE_EMPTY_INDEX, 13 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_CONTENT_TYPE_EMPTY_INDEX, 12 + 32, GRPC_BATCH_CONTENT_TYPE},
+    {GRPC_MDELEM_COOKIE_EMPTY_INDEX, 6 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_DATE_EMPTY_INDEX, 4 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_ETAG_EMPTY_INDEX, 4 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_EXPECT_EMPTY_INDEX, 6 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_EXPIRES_EMPTY_INDEX, 7 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_FROM_EMPTY_INDEX, 4 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_HOST_EMPTY_INDEX, 4 + 32, GRPC_BATCH_HOST},
+    {GRPC_MDELEM_IF_MATCH_EMPTY_INDEX, 8 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_IF_MODIFIED_SINCE_EMPTY_INDEX, 17 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_IF_NONE_MATCH_EMPTY_INDEX, 13 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_IF_RANGE_EMPTY_INDEX, 8 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_IF_UNMODIFIED_SINCE_EMPTY_INDEX, 19 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_LAST_MODIFIED_EMPTY_INDEX, 13 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_LINK_EMPTY_INDEX, 4 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_LOCATION_EMPTY_INDEX, 8 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_MAX_FORWARDS_EMPTY_INDEX, 12 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_PROXY_AUTHENTICATE_EMPTY_INDEX, 18 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_PROXY_AUTHORIZATION_EMPTY_INDEX, 19 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_RANGE_EMPTY_INDEX, 5 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_REFERER_EMPTY_INDEX, 7 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_REFRESH_EMPTY_INDEX, 7 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_RETRY_AFTER_EMPTY_INDEX, 11 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_SERVER_EMPTY_INDEX, 6 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_COOKIE_EMPTY_INDEX, 10 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_STRICT_TRANSPORT_SECURITY_EMPTY_INDEX, 25 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_TRANSFER_ENCODING_EMPTY_INDEX, 17 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_USER_AGENT_EMPTY_INDEX, 10 + 32, GRPC_BATCH_USER_AGENT},
+    {GRPC_MDELEM_VARY_EMPTY_INDEX, 4 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_VIA_EMPTY_INDEX, 3 + 32, GRPC_BATCH_CALLOUTS_COUNT},
+    {GRPC_MDELEM_WWW_AUTHENTICATE_EMPTY_INDEX, 16 + 32,
+     GRPC_BATCH_CALLOUTS_COUNT},
+};
+
+/* This is a faster check for seeing if a mdelem index is used or not. To verify
+   that the index value is valid, use 'is_valid_mdelem_index' */
+static bool is_mdelem_index_used(uint8_t index);
```

Comment:
I'm aware that this function, as well as set_mdelem_index_unused, should probably take a linked_mdelem instead of an uint8_t or a uint8_t*. I'm also aware that in the places where I'm setting the linked_mdelem's index to 0, I should really be calling `set_mdelem_index_unused`. I'll make these changes in a later PR because the tests have been flaky/difficult and I'd like to get this in.